<center><img src="https://i.imgur.com/zRrFdsf.png" width="700"></center>










# Multivariate Geostats: Beyond classical Clustering and Regression

## Getting Ready

### The Data


Let's use the data from Peru previously seen:

In [ ]:
import geopandas as gpd

peruDataLink="https://github.com/CienciaDeDatosEspacial/dataSets/raw/refs/heads/main/PERU/PeruMaps.gpkg"
DisMap=gpd.read_file(peruDataLink,layer='good_geom')

The "good_geom" layer is a new layer we had to use because it show no relevant topological issues (we had previously used one with some issues not detected).

Rememeber this "good map" has isolates:

In [ ]:
from libpysal.graph import Graph

DisMap_adjaceny=Graph.build_contiguity(DisMap)
DisMap_adjaceny.isolates

Let's subset now, and avoid future warnings:

In [ ]:
DisMap.drop(index=DisMap_adjaceny.isolates,inplace=True)
DisMap.reset_index(drop=True,inplace=True)

### Preprocessing

We have these columns:

In [ ]:
DisMap.columns

Let me select the columns we will use:

In [ ]:
selected_variables = ['Educ_sec_comp2019_pct','NBI2017_pct', 'Viv_sin_serv_hig2017_pct']

Let's see the descriptives, paying attention to **range**:

In [ ]:
DisMap[selected_variables].describe()

Let me check their monotony:

In [ ]:
DisMap[selected_variables].corr()

Clearly, 'Educ_sec_comp2019_pct' represents a diffent monotonic direction than the rest, because it measures a positive concept, while the others measure negative ones. 

Here, we can reverse the values of *Educ_sec_comp2019_pct*:

In [ ]:
## from complete to 'incomplete'
DisMap['Educ_sec_incomp2019_pct']=100-DisMap['Educ_sec_comp2019_pct']

Let's verify the presence of outliers:

In [ ]:
DisMap[selected_variables].plot(kind='box')

The presence of outliers forces us to choose a robust transformation to produce columns ready for clustering techniques:

$$Z_{\text{robust}} = \frac{X - \text{Median}}{\text{MAD}}$$

1.  **Median:** Unlike the mean, the median is minimally affected by outliers.
2.  **Median Absolute Deviation (MAD):** This is a robust measure of statistical dispersion. It is calculated by finding the median of the absolute deviations from the median: $\text{MAD} = \text{Median}(|X - \text{Median}|)$.

I will use **mad()** from scipy with **scale='normal'**. The `scale='normal'` argument ensures the resulting MAD value is scaled to be comparable to the standard deviation ($\sigma$). This ensures your new robust scores are on a familiar scale (most values between $-3$ and $+3$) and are suitable for our clustering algorithm.

Notice regression **needs not** this standardization.

In [ ]:
import pandas as pd
import numpy as np
from scipy.stats import median_abs_deviation as mad

# --- Function to Perform Robust Standardization (Z-score via Median and MAD) ---
def robust_score(series):
    # 1. Calculate the Median
    median_val = series.median()
    # 2. Calculate the Median Absolute Deviation (MAD)
    mad_val = mad(series, scale='normal') #for comparability with Zs
    # 3. Apply the Robust Formula
    return (series - median_val) / mad_val

Let's rename the variables to used during clustering:

In [ ]:
deficit_cols = ['Educ_sec_incomp2019_pct','NBI2017_pct', 'Viv_sin_serv_hig2017_pct']

clusterNames=['Education_deficit','BasicNeeds_deficit','Sanitation_deficit']
DisMap[clusterNames]=DisMap[deficit_cols].apply(robust_score)
DisMap[clusterNames]

In [ ]:
# as a result:
DisMap[clusterNames].corr()

Let's rename the original variables for regression section:

In [ ]:
#prepare the dict
regressionNames=[col+'_pct' for col in clusterNames]
changes={old:new for old,new in zip(deficit_cols,regressionNames)}
changes

In [ ]:
DisMap.rename(columns=changes,inplace=True)

### The SEED

Several functions will use some random process. I will set a seed here, that you have use should you wnat to get the same results I got.

In [ ]:
import numpy
RANDOM_SEED = 12345
numpy.random.seed(RANDOM_SEED)

## Clustering

### Conventional Clustering

Conventional clustering is a basic machine learning technique used to group a set of data points into clusters so that points within the same cluster are more similar to each other than to those in other clusters. Think of it as sorting objects into piles based on how alike they are.

In this case, I will use the three variables to create clusters of **districts**. There are several options you may have seen before, like KMeans,KMedoids, and Hierarchical (divissive/agglomerative). 


Let me use **KMeans**. This technique groups data points into **K number of clusters** by making the points in each cluster as similar as possible. You may have seen this in a previous course, but the simpler way to depict the process might be:

> Imagine you have a large box filled with mixed-up balls of various colors. To sort them into three clusters, you start by randomly picking three balls to act as the initial group leaders. As you pick more balls, you assign each one to the group whose leader has the most similar color. Periodically, you re-evaluate and choose new, better leaders for the groups based on the balls you've already sorted. This process repeats until, eventually, you find that the balls have separated into distinct groups of bluish, reddish, and greenish balls.

As you see, the algorithm is a process of **refinement**. It starts with a random guess and keeps improving the groups until it finds the most compact, natural-looking clusters possible.


| Analogy Above | Formal K-Means Step | What Happens |
| :--- | :--- | :--- |
| **"I want 3 clusters"** | **1. Choose K** | Decide on the number of clusters. |
| **"Pick 3 random balls as leaders"** | **2. Initialize** | Place K random centroids in the data. |
| **"Balls gather with the leader they look most similar to"** | **3. Assign** | Assign each point to its nearest centroid, forming clusters. |
| **"Find a better color and change the leader"** | **4. Update** | Calculate the new centroid (mean) for each cluster. |
| **"That will happen again and again..."** | **5. Repeat** | Re-assign and update until centroids stabilize. |

K-Means is an **unsupervised learning** algorithm used to find natural groupings in unlabeled data. It's great for exploration and pattern discovery.

> **Remember:** You have to tell it how many clusters (K) you want in advance. 

In [ ]:
from sklearn.cluster import KMeans

K_target = 5

# as simple as:
DisMap['deficit_k5'] = KMeans(n_clusters=K_target).fit_predict(DisMap[clusterNames])

In [ ]:
# see distribution of districts
DisMap["deficit_k5"].value_counts()

We could try to find the pattern that created the clusters:

In [ ]:
DisMap.groupby("deficit_k5")[clusterNames].mean()

Keepig in mind that meaning a score of **0** is the national median, a **positive** score indicates a **higher deficit (worse)** than the median, and a **negative** score indicates a **lower deficit (better)** than the median; let's propose an interpretation, and naming, of our five distinct clusters:


| Cluster | Label | Reason |
|---------|-------|---------|
| **0** | `Low Deficits - Well Performing` | All deficits negative (below average) |
| **1** | `Sanitation Crisis Area` | Extreme sanitation deficit (+1.46) |
| **2** | `Severe Multi-Dimensional Crisis` | All deficits highly positive (worst) |
| **3** | `Education Challenge Area` | Only education deficit is positive |
| **4** | `Education & Basic Needs Struggle` | High education & basic needs deficits |

**Quick interpretation:**
- **Cluster 0**: Best areas (intervention not urgent)
- **Cluster 2**: Worst areas (need immediate comprehensive intervention)  
- **Clusters 1,3,4**: Need targeted interventions for specific deficits


## Regionalization: Spatial Clustering 

Regionalization is the process of **clustering cases into contiguous regions** based on shared characteristics. In our work, the goal is to create socio-economic zones where districts within a zone are similar to each other and geographically connected.

Conventional clustering (like KMeans above) group districts based only on their **deficit scores** (Education, NBI, Sanitation), but you see  a messy, fragmented map.  **Spatial Clustering (Regionalization) will wotk differently following:**

1.  **Attribute Similarity:** It still requires districts to have similar socio-economic **deficits** (based on your Robust Z-scores).
2.  **Contiguity Constraint:** It adds the rule that a district can **only be grouped with a neighbor it physically shares a border with**.

By enforcing this contiguity, spatial clustering ensures that the final $K=5$ groups  are not scattered patches but are **contiguous, policy-relevant territorial blocks**.

This way Spatial clustering tries building regions that are both **statistically coherent** (similar deficit profiles) and **geographically coherent** (connected in space). A good regionalization  makes the results useful for regional planning, resource allocation, and targeted policy intervention. 



Let's compute the neigborhood matrix:

In [ ]:
# for mos cases except spatial kmeans
peru_perim_graph = Graph.build_contiguity(DisMap, by_perimeter=True)
peru_perim_graph=peru_perim_graph.transform("r")

# for spatial kmeans
from libpysal.weights import W # Import W for the conversion
peru_perim = peru_perim_graph.to_W()
peru_perim.transform = "r"


### 1. Spatial KMeans


**Input:**
- Geographic coordinates of each unit
- The number of regions (K) needed
- Attribute data for clustering

**Process:**
Iterative refinement combining spatial and attribute similarity:
- Start with K random seed locations as initial region centers
- In each iteration:
  - Assign each unit to the closest region center considering both:
    - Spatial proximity (geographic distance)
    - Attribute similarity (feature space distance)
  - Update region centers based on current assignments
- Continue until assignments stabilize

**Output:**
- Geographic units partitioned into K regions
- Regions balance spatial compactness with attribute homogeneity


We can easily get these clusters using:

In [ ]:
from spopt.region import RegionKMeansHeuristic as RKMeans

RKMeans_model = RKMeans(DisMap[clusterNames].values, K_target, w=peru_perim)
# computing
RKMeans_model.solve()

Now, a new column with the cluster labels.

In [ ]:
DisMap['deficit_k5R'] = RKMeans_model.labels_

### 2. AZP (A Priori Zoning Problem)

AZP is one of the earliest and most direct algorithms for spatial clustering.

**Input:**
- A graph representing neighborhood relationships
- The number of regions (K) needed
- An initial random partition of regions

**Process:**
Iteratively improve region assignments through "node swapping" where:
- Each geographic unit considers moving to adjacent regions
- Moves are accepted if they improve regional homogeneity
- The algorithm seeks to minimize within-region dissimilarity
- Continues until no improving moves can be found

**Output:**
- Geographic units partitioned into K spatially contiguous regions
- Regions are locally optimal in terms of internal homogeneity


In [ ]:
from spopt.region import AZP

azp_model = AZP(
    gdf=DisMap,
    w=peru_perim_graph,
    attrs_name=clusterNames,
    n_clusters=K_target,
    random_state=RANDOM_SEED)
# computing
azp_model.solve()

New column with labels:

In [ ]:
DisMap['deficit_azpR'] = azp_model.labels_

### 3. Skater (Spatial 'K'luster Analysis by Tree Edge Removal)


**Input:**
- A graph representing neighborhood relationships  
- The number of regions (K) needed

**Process:**
Transform the neighborhood graph into a Minimum Spanning Tree ([MST](https://colab.research.google.com/drive/1CQwjJR6i_3X_8izLsDt-TPfbLtpCnN5T?usp=sharing)) where:
- Each node connects to its most similar neighbors
- Similarity is computed from social variables used for clustering  
- The MST minimizes total dissimilarity while maintaining full connectivity

**Output:**
- Geographic units partitioned into K spatially contiguous regions
- Each region is internally homogeneous and geographically connected

In [ ]:
from spopt.region import Skater

skater_model = Skater(
    gdf=DisMap,
    w=peru_perim_graph,
    attrs_name=clusterNames,
    n_clusters=K_target
)
# computing
skater_model.solve()


The cluster as a new column:

In [ ]:
DisMap['deficit_skaterR'] = skater_model.labels_

## Evaluating regionalization

### COMPACTNESS

* **The isoperimetric quotient (IPQ)**. This compares the area of the region to the area of a circle with the same perimeter as the region.

In [ ]:
from esda import shape as compactness

# columns with cluster labels
cluster_types_to_evaluate = ["deficit_k5","deficit_k5R","deficit_azpR","deficit_skaterR"]

results = {}

for cluster_type in cluster_types_to_evaluate:
    ## Dissolve the units into contiguous regions based on the cluster label
    regions = DisMap[[cluster_type, "geometry"]].dissolve(by=cluster_type)
    
    ## Compute the IPQ for the dissolved regions
    ipqs = compactness.isoperimetric_quotient(regions)
    
    ## Store the results
    results[cluster_type] = ipqs

# Create the final comparison DataFrame
ipq_comparison_df = pd.DataFrame(results)

# Adding the mean IPQ for each model
ipq_comparison_df.loc['Mean_IPQ'] = ipq_comparison_df.mean()

ipq_comparison_df

* **Convex hull ratio (CHR)**. The CHR is simply the division of the area of the cluster by the area of its convex hull.

In [ ]:
from esda import shape as compactness
results={}
for cluster_type in cluster_types_to_evaluate:
    ## Dissolve the units into contiguous regions based on the cluster label
    regions = DisMap[[cluster_type, "geometry"]].dissolve(by=cluster_type)
    ## Compute the CHR for the dissolved regions
    chullr = compactness.convex_hull_ratio(regions)
    ## Store the results
    results[cluster_type] = chullr

# Create the final comparison DataFrame
chr_comparison_df = pd.DataFrame(results)

# Adding the mean IPQ for each model
chr_comparison_df.loc['Mean_CHR'] = chr_comparison_df.mean()

# see
chr_comparison_df

Both **Isoperimetric Quotient (IPQ)** and **Convex Hull Ratio (CHR)** are metrics used to assess the **spatial compactness** of a polygon (like a geographic region), but they quantify compactness based on different geometric ideals.

Here is a breakdown of their differences, how they are calculated, and what they measure:


| Feature | Isoperimetric Quotient (IPQ) | Convex Hull Ratio (CHR) |
| :--- | :--- | :--- |
| **Ideal Shape** | **Circle** (the most compact shape) | **Any Convex Shape** (the simplest possible enclosure) |
| **Formula** | $$\text{IPQ} = \frac{4\pi \cdot \text{Area}}{\text{Perimeter}^2}$$|$$\text{CHR} = \frac{\text{Area of the Region}}{\text{Area of the Convex Hull}}$$ |
| **Range** | $0$ (least compact) to **$1.0$** (perfect circle) | $0$ (least compact) to **$1.0$** (perfectly convex) |
| **Focus** | **Boundary Complexity/Elongation:** It penalizes long, wiggly, or highly elongated shapes by squaring the perimeter. | **Boundary Indentation/Concavity:** It penalizes "pitted" or "re-entrant" shapes where the boundary folds inward. |
| **Use Case** | Excellent for assessing **efficiency** and **elongation** (e.g., in electoral districts). | Excellent for assessing **irregularity** and **concavity** (e.g., in regionalization studies). |


### HOMOGENEITY

Homogeneity, in the context of spatial clustering and regionalization (like with PySAL), refers to the **degree of similarity** among the smaller geographic units *within* a newly formed, larger region.

Simply put, it measures **how alike the members of a group are.**

Homogeneity is the **primary objective** of spatial clustering algorithms like RegionKMeans, AZP, and SKATER. These methods are designed to create regions where the difference in the chosen attributes (features) is minimized internally, even while adhering to the constraint of spatial contiguity.

Homogeneity is assessed using the **attribute data** ($\mathbf{X}$) used for clustering (our _deficit variables_). Homogeneity is mathematically quantified using measures of **dispersion** or **variance** within each cluster. Let's compute the  **Calinski-Harabasz (CH) Score** ; and **Silhouette (SIL) Score**:

In [ ]:
from sklearn import metrics

fit_scores = []

for cluster_type in cluster_types_to_evaluate:
    ch_score = metrics.calinski_harabasz_score(
        DisMap[clusterNames],# using scaled variables
        DisMap[cluster_type])# using these labels
    sil_score = metrics.silhouette_score(
        DisMap[clusterNames],DisMap[cluster_type]) #same as above
    
    # appending 
    fit_scores.append((cluster_type, ch_score,sil_score))


# re-arrange the scores into a dataframe for display
colNames=["cluster type", "CH score", "SIL score"]
pd.DataFrame(fit_scores, columns=colNames).set_index("cluster type")

Here is a comparison table outlining the core differences between the two metrics:

| Feature | Silhouette Score (SIL) | Calinski-Harabasz (CH) Score |
| :--- | :--- | :--- |
| **Calculation Base** | **Point-to-Point Distance** (Compares distance to own cluster vs. next-nearest cluster) | **Variance/Centroid Distance** (Ratio of between-cluster variance (SSB) to within-cluster variance (SSW)) |
| **Range** | **$-1$ to $+1$** | $\ge 0$ (Unbounded) |
| **Interpretation** | Closer to **+1** means high cohesion and high separation. | Higher values mean high cohesion and high separation. |
| **Best Use** | Assessing **cluster tightness** and identifying poorly classified/ambiguous points. | Quickly identifying the **optimal number of clusters** ($k$). |


Our table compares the homogeneity and separation of four different regionalization solutions for the `deficit` variable.

| Metric | cluster type |
| :--- | :--- |
| **Benchmark (Non-Spatial)** |  `deficit_k5` |
| **Spatial 1 (RegionKMeans)** |  `deficit_k5R` |
| **Spatial 2 (AZP)** | `deficit_azpR` |
| **Spatial 3 (SKATER)** | `deficit_skaterR` |

The results show a stark contrast between the **non-spatial K-means benchmark** and the three **spatially constrained regionalization methods**:

* **Massive Drop in CH Score in all Spatial techniques:** The spatial constraint (forcing regions to be contiguous) introduces a **significant cost** in terms of attribute homogeneity. The algorithms have to group adjacent units, even if they are very different in terms of `deficit` attributes, leading to much lower separation (SSB) and higher internal variance (SSW). This is normal and expected when moving from non-spatial to spatial clustering.

* **Positive vs. Negative SIL Score:** The non-spatial K-means (`deficit_k5`) has a moderately good (positive) SIL score, while, **all three spatial solutions** have **negative** SIL scores. A negative Silhouette Score means that, on average, the units are **closer to the centroid of another region** (their neighbor) than they are to their own region's centroid. This reinforces the finding that the spatial constraint is forcing highly non-homogeneous groupings. 

Comparing the three spatial methods that adhere to the contiguity constraint:

* **Best Homogeneity/Separation (Highest CH):** **AZP** (`deficit_azpR`) got the highest  CH score. This suggests that, among the constrained solutions, AZP found the regional configuration that best balanced internal similarity and external separation.
* **Worst Homogeneity/Separation (Lowest SIL):** **SKATER** (`deficit_skaterR`) has the lowest SIL score. This indicates that SKATER produced the most ambiguous and least cohesive regions, relative to the two k-means-based methods (RegionKMeans and AZP).
* **Regional-k-means (`deficit_k5R`):** This solution falls between AZP and SKATER in quality.

Based purely on homogeneity and separation metrics, **AZP is the best performing regionalization method** for the `deficit` variable, although all spatial solutions show a substantial penalty for enforcing spatial contiguity.

In summary,


**Choose AZP if... (The Statistical Path)** if the primary goal is to **model or explain** the 'deficit' attribute using the new regional structure.

* **Policy Analysis:** You are trying to understand the spatial distribution and clustering of the deficit for academic research or modeling purposes.
* **Predictive Modeling:** The region IDs will be used as a variable in a regression or machine learning model, and you need the strongest possible signal (least internal noise) from your regional groups.
* **Targeting:** You need to identify areas where the average characteristics are most uniform to ensure any intervention is perfectly tailored to the demographic/economic reality of that region.

**Choose SKATER if... (The Practical Path)** if the results will be used to create **real-world administrative or political boundaries** where shape and simplicity matter most.

* **Administrative/Service Areas:** The regions will define boundaries for resource allocation, local governance, or service delivery (e.g., school districts, health zones). Irregular shapes lead to inefficient routes and unequal service distribution.
* **Electoral Redistricting:** Compactness and simplicity are often legal or ethical requirements in defining political districts to prevent gerrymandering. 
* **Visualization:** You need a map that is clean, easy to interpret, and visually representative without highly complex, interdigitated boundaries.

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

# Assuming DisMap is your GeoDataFrame with the cluster columns
cluster_types = ["deficit_k5", "deficit_k5R", "deficit_azpR", "deficit_skaterR"]
titles = ["K-Means", "RegionKMeans", "AZP", "SKATER"]

fig, axes = plt.subplots(2, 2, figsize=(16, 12))
axes = axes.flatten()

# Find the maximum number of clusters across all methods for consistent coloring
max_clusters = max([DisMap[ct].nunique() for ct in cluster_types])

for i, (cluster_type, title) in enumerate(zip(cluster_types, titles)):
    # Plot with consistent colormap
    DisMap.plot(column=cluster_type, 
                ax=axes[i],
                categorical=True,
                legend=True,
                cmap='tab10',  # Consistent colors
                edgecolor=None,
                linewidth=0.3)
    
    axes[i].set_title(f'{title}', fontsize=14, fontweight='bold')
    axes[i].set_axis_off()

plt.suptitle('Comparison when K is 5\n (color can not be compared without profiling)', 
             fontsize=16, fontweight='bold',color='blue')
plt.tight_layout()
plt.show()




## Conventional Regression


Let's represent this model:

$$\text{BasicNeeds\_deficit\_pct} = \beta_0 + \beta_1 (\text{Sanitation\_deficit\_pct}) + \beta_2 (\text{Education\_deficit\_pct}) + \epsilon$$

The model attempts to predict the **Percentage of Population with Unmet Basic Needs (NBI)** using the **Percentage with Completed High-School Education** and the **Percentage of Households Lacking Sanitation**.


In [30]:
from pysal.model import spreg

dep_var_name='BasicNeeds_deficit_pct'
ind_vars_names=['Sanitation_deficit_pct', 'Education_deficit_pct']

argsRegre=dict(
    y=DisMap[dep_var_name], # Dependent variable
    x=DisMap[ind_vars_names], # Independent variables
    name_y=dep_var_name, # Dependent variable name
    name_x=ind_vars_names, # Independent variable name
    name_ds='DisMap') # name of data set

ols_model = spreg.OLS(**argsRegre)
    

ols_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,11.230543,0.876661,12.810583,0.0
1,Sanitation_deficit_pct,0.671289,0.018089,37.110163,0.0
2,Education_deficit_pct,0.359272,0.018809,19.100679,0.0


If we wish to compare this output with other regression models, you may use:

* For **non-nested models**:

In [31]:
ols_model.ar2

np.float64(0.5301902863255213)

In [32]:
ols_model.aic

np.float64(14775.887907529901)

In [33]:
ols_model.logll # Log Likelihood

np.float64(-7384.9439537649505)

In [34]:
np.sqrt(ols_model.sig2) # S.E. of Regression

np.float64(12.539968014149354)

* For **nested models**:

In [35]:
ols_model.f_stat[0],ols_model.f_stat[1]<0.05

(np.float64(1056.167450321906), np.True_)

| Test | Null Hypothesis ($H_0$) | Comparison Method | Key Concept |
| :--- | :--- | :--- | :--- |
| **F-statistic** | All $X$ coefficients are zero (comparing the model to a null model with only a constant). | Reject $H_0$ if **$p$-value** is low. | **Overall Model Significance** |

In simple terms, when comparing different models:

> The model with the **lowest AIC**, the **highest Adjusted $R^2$**, and the **lowest S.E. of Regression** is generally preferred, given the **F-Statistics is significant**.


### OLS diagnostics

WE need to verify post-hoc if some issues arised that would weaken model usefulness.

The object **ols_model** has key information on this:

In [38]:
# is there collinearity?
ols_model.mulColli>=30

False

In [42]:
# is the normality?
ols_model.jarque_bera['pvalue']>0.05 

np.False_

In [43]:
# is there homocedasticity?
ols_model.koenker_bassett['pvalue']>0.05

np.False_

If we stay with conventional regression (OLS), we should deal with lack of normality and homocedasticity. 


### Spatial Regression need?

Let's re run the previous code, but requesting **spatial disgnostics**, for that we input the adjacency matrix (peru_perim_graph)

In [47]:
W_info=dict(
    w=peru_perim_graph,       # REQUIRED: Defines neighbor relationships for spatial diagnostics
    name_w='peru_perim_graph') # Name of the weights matrix

argsRegre_W=argsRegre.copy()
argsRegre_W.update(W_info)

ols_model_with_W = spreg.OLS( **argsRegre_W,                 
                   spat_diag=True,         # generates spatial tests)
                   robust='white'          # essential to deal with OLS diagnostic problems
                   ) 

Given we request spatial diagnostics (post-hoc), let's apply the **Durbin Joint Test**:

In [49]:
# Do any of the regression terms (X, Y, error) is being affected by behavior of neighbors
ols_model_with_W.lm_spdurbin[1]<0.05

np.True_

The significance of the Joint test for SDM (your result: $p < 0.05$) confirms a crucial point:

>The OLS model is statistically wrong (misspecified). A spatial model is needed. This is telling us that location is playing in at least one of these terms: X (independent variables), Y (dependent variables), $\epsilon$ (error term).

These are possible regression models for each case:


* **SAR**: The value of Y in neighboring locations directly affects your Y.

In [63]:
ols_model_with_W.rlm_lag[1]<0.05

np.False_

* **SEM**: A location's error is dependent on its neighbors' errors, meaning unobserved influences spread across space

In [64]:
ols_model_with_W.rlm_error[1]<0.05

np.True_

* **SLX**: The values of X in neighboring locations affect your Y.

In [66]:
ols_model_with_W.rlm_wx[1]<0.05

np.True_

Let's start the journey! (SAR was not good, but will be part of the journey).

## Spatial Regression: SAR

The Spatial Autoregressive (SAR) model can be computed like this:


In [67]:
SAR_model = spreg.GM_Lag(
    **argsRegre_W,
    spat_impacts='simple', 
    robust='white') # we knew there is Heteroskedasticity

Let's check the GoF first:

In [59]:
# potentially better than OLS
ols_model.ar2<SAR_model.pr2 #pseudo R2

np.True_

Now, let's see the regression coefficients:

In [60]:
SAR_model.output # this is rho: W_BasicNeeds_deficit_pct

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,9.767108,1.048124,9.318655,0.0
1,Sanitation_deficit_pct,0.653007,0.018482,35.331866,0.0
2,Education_deficit_pct,0.33859,0.023903,14.165268,0.0
3,W_BasicNeeds_deficit_pct,0.069026,0.035413,1.949188,0.051273


The **W\_BasicNeeds\_deficit\_pct**, is the estimated **Spatial Autoregressive Parameter ($\mathbf{\rho}$ or *rho*)** for the SAR model. This $\mathbf{\rho}$ captures the existence of **Spillover Effect**:

> **POSITIVE SPILLOVER:** The positive value ($\mathbf{0.069}$) indicates **positive spatial autocorrelation** in the dependent variable. That is, an increase in the average `BasicNeeds_deficit_pct` in neighboring 'distrito' leads to an increase in the `BasicNeeds_deficit_pct` in the current 'distrito', holding all other factors constant. The magnitude is relatively small ($\mathbf{0.069}$), meaning the spillover effect is not very strong. Morevover, the **$p$-value:** $\mathbf{0.051273}$ is just barely above the conventional $0.05$ significance level.

At a strict $\alpha=0.05$ level, you would **fail to reject the null hypothesis ($H_0: \rho = 0$)**, suggesting the spatial lag term is **not statistically significant**.

* **Diagnostics for remaining Spatial Dependence**:

In [81]:
#Anselin-Kelejian Test
SAR_model.ak_test, SAR_model.ak_test[1]<0.05

((np.float64(132.56784792511712), np.float64(1.1239891603541386e-30)),
 np.True_)

Then,

> **The SAR model has NOT fully corrected the spatial dependence problem, as the model residuals are still spatially correlated.**


* **Interpretation of coefficients**

In a SAR model, the coefficients ($\beta$) are not direct marginal effects because a change in $X$ in one location affects the dependent variable in all other locations (due to the feedback loop of $\rho WY$). Therefore, the impacts are broken down:

In [71]:
from spreg.sputils import spmultiplier

# Compute multipliers
impacts = spmultiplier(peru_perim_graph, SAR_model.rho[0])

# Get relevant betas (drop intercept and rho)
betas = SAR_model.betas[1:-1].flatten()

# Build impact table
effects_df = pd.DataFrame({
    'Variable': ind_vars_names,
    'Direct': betas * impacts['adi'],
    'Indirect': betas * impacts['aii'],
    'Total': betas * impacts['ati']
})

# here you are
effects_df


,Variable,Direct,Indirect,Total
0,Sanitation_deficit_pct,0.653007,0.048416,0.701423
1,Education_deficit_pct,0.338590,0.025104,0.363694


In our case, the **SAR** model assumes that the level of unsatisfied needs in each city depends not only on its own socioeconomic deficits, but also on the **levels of unsatisfied needs in neighboring cities** (through ( WY ) and ρ).
 
Because ρ = 0.069 (small but positive), there **might be some spatial dependence**, meaning deprivation in one city tends to “echo” in its neighbors — but modestly.

**Sanitation_deficit_pct**

* **Direct effect (0.653):**
  If the share of households without adequate sanitation in a city increases by **1 percentage point**, the share of the population with *unsatisfied needs* in that city increases by **0.65 percentage points**, on average.
  → This captures the *within-city* marginal impact.

* **Indirect effect (0.048):**
  The same 1-point increase also raises unsatisfied needs in *neighboring cities* by about **0.05 percentage points**, on average, due to spatial spillovers.
  → Poor sanitation in one city slightly worsens deprivation nearby (perhaps via migration, shared infrastructure, or regional contagion of poverty).

* **Total effect (0.701):**
  Overall, each additional point of sanitation deficit in one city leads to about **0.7 percentage points more unsatisfied needs** in the regional system (the city + its neighbors).


**Education_deficit_pct**

* **Direct effect (0.339):**
  A 1-point increase in the local education deficit raises unsatisfied needs locally by **0.34 points**, on average.

* **Indirect effect (0.025):**
  It also raises deprivation in nearby cities by **0.025 points**, on average.

* **Total effect (0.364):**
  Overall regional effect ≈ **0.36 percentage points** increase in unsatisfied needs for each point increase in education deficit.

What the pattern tells us:

* **Local impacts dominate:** Most of the effect of social deficits operates **within each city**, as shown by the ratio of Direct to Total effects (≈ 92–93 % for both variables). This indicates that the bulk of deprivation is locally determined, with limited interdependence among neighboring cities.

* **Spillovers are small but systematic:** The remaining 7–8 % of the total effect reflects **spatial spillovers**, meaning that a city’s deficits slightly exacerbate unsatisfied needs in adjacent areas. This pattern is consistent with **spatial diffusion of poverty** and the emergence of **regional inequality traps**, where deprivation in one locality reinforces that of nearby ones.

* **Comparatively:** **Sanitation deficits** have roughly **twice the total impact** of education deficits on unsatisfied needs (0.70 vs. 0.36). This suggests that deficits in **basic infrastructure and public services** play a more decisive role in shaping aggregate social deprivation than shortfalls in education.

## Spatial Regression: SEM


The **Spatial Error Model (SEM)** can be computed like this:


In [80]:
# GM - uses a Generalized Moments (GM) estimator that is robust to heteroskedasticity.
SEM_model = spreg.GM_Error_Het(**argsRegre_W)

Let's check the GoF first:

In [76]:
SEM_model.pr2>ols_model.ar2

np.False_

Now the regression coefficients:

In [79]:
SEM_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,16.291147,1.150997,14.153945,0.0
1,Sanitation_deficit_pct,0.744934,0.014981,49.726259,0.0
2,Education_deficit_pct,0.203697,0.021447,9.497633,0.0
3,lambda,0.780674,0.018528,42.13485,0.0


The estimated spatial error coefficient of $\lambda = \mathbf{0.781}$ indicates a very strong spatial dependence in the unobserved component of the model. Substantively, this means that an unexplained shock in one district induces, on average, a response equal to about 78% of that shock’s magnitude in the composite error of its neighboring districts, as defined by the spatial weights matrix W. In other words, local unobservables are highly correlated across space: when one area experiences higher-than-predicted Basic Needs Deficit, nearby areas tend to exhibit nearly proportional residual increases. Such a large and significant value confirms that the residual process is strongly autoregressive, and that controlling for spatial dependence through the SEM specification is essential.

* **Heteroskedasticity-Robust Coefficients**

The standard errors and z-statistics here are corrected for both spatial error and heteroskedasticity, making them the **most reliable estimates so far**.

## Spatial Regression: SLX

The Spatial Lag of X (SLX) model cam be computed like this:

In [87]:
slx_model = spreg.OLS(**argsRegre_W,
    spat_diag=True,
    slx_lags=1,robust="white"
)
slx_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,7.200011,1.017128,7.078765,0.0
1,Sanitation_deficit_pct,0.724735,0.020432,35.471414,0.0
2,Education_deficit_pct,0.203485,0.030228,6.73176,0.0
3,W_Sanitation_deficit_pct,-0.112312,0.031362,-3.581161,0.000351
4,W_Education_deficit_pct,0.269385,0.037319,7.218526,0.0


Given that SLX models:


$$
y = X\beta + WX\theta + \varepsilon
$$

* **Direct effect** → the impact of a variable in district *i* on its own outcome (*β*).
* **Indirect effect (spillover)** → the impact of that variable in *neighboring districts* on district *i*’s outcome (*θ*).
* **Total effect** → the sum (*β + θ*), representing the **overall regional impact** of a change in that variable.


| Variable               | Direct effect (β) | Indirect effect (θ) | Total effect (β + θ) | Significance  |
| ---------------------- | ----------------: | ------------------: | -------------------: | ------------- |
| **Sanitation deficit** |            0.7247 |             −0.1123 |           **0.6124** | All p < 0.001 |
| **Education deficit**  |            0.2035 |             +0.2694 |           **0.4729** | All p < 0.001 |

We see that:

> **Direct effects** show that both sanitation and education deficits significantly increase local deprivation: a one-unit rise in sanitation deficit raises Basic Needs Deficit by **0.725 points**, while a comparable increase in education deficit raises it by **0.203 points**.
>
> **Indirect (spillover) effects** indicate that neighboring conditions also matter. Higher **education deficits in adjacent districts** raise local deprivation (+0.269), revealing a **positive spatial spillover**—deprivation clusters regionally through educational disadvantage. Conversely, **neighboring sanitation deficits** exhibit a **negative spillover** (−0.112), suggesting **spatial compensation or policy substitution**, where areas surrounded by worse sanitation conditions may receive offsetting resources or attention.
>
> **Total effects** confirm that sanitation deficits remain the strongest overall determinant (0.612), but **educational deficits—local and neighboring combined—exert nearly half as much total impact (0.473)**, underscoring the regional diffusion of educational deprivation.



* Are we missing something given we have controlled spatial dependence in X?

In [96]:
# need to further control error?
slx_model.rlm_error[1]<0.05

np.False_

In [95]:
# need to further control Y?
slx_model.rlm_lag[1]<0.05

np.True_

## Spatial Regression: SLX-Error


The Spatially Lagged X with Spatial Error Model (**SLX-Error Model**) is a hybrid spatial regression model designed to account for two distinct sources of spatial dependence simultaneously:

1.  **Exogenous Spatial Interaction (The SLX Component):** The influence of independent variables ($\mathbf{X}$) in neighboring areas directly affecting the dependent variable ($\mathbf{Y}$) in the local area.
2.  **Unmodeled Spatial Dependence (The Error Component):** The spatial clustering of unobserved factors that are not captured by the independent variables, which shows up in the error term.

The SLX-Error model essentially takes the standard **Spatial Error Model (SEM)** and adds the **Spatially Lagged X ($\mathbf{W X}$) terms**.

The SLX-Error model is defined by the following two equations:

$$
\begin{aligned}
\mathbf{Y} &= \mathbf{X}\boldsymbol{\beta} + \mathbf{W X}\boldsymbol{\theta} + \boldsymbol{\mu} \\
\boldsymbol{\mu} &= \lambda \mathbf{W}\boldsymbol{\epsilon} + \boldsymbol{\epsilon}
\end{aligned}
$$


| Term | Description | Role in the Model |
| :--- | :--- | :--- |
| $\mathbf{W X}\boldsymbol{\theta}$ | **Spatially Lagged Independent Variables (SLX)** | Captures **exogenous spillovers**. For example, your basic needs deficit is affected by your neighbor's sanitation deficit. |
| $\lambda \mathbf{W}\boldsymbol{\epsilon}$ | **Spatially Autocorrelated Error Term** | Captures **unobserved factors** that are spatially clustered (e.g., shared unmeasured infrastructure or regional policy effects). |

Key Advantages and Interpretation:

* **Exogenous Spillover:** The $\mathbf{W X}\boldsymbol{\theta}$ terms allow for the direct quantification and interpretation of **exogenous spillovers**. The coefficient $\boldsymbol{\theta}$ is interpreted directly: a change in a neighbor's $X$ leads to a $\theta$-unit change in your $Y$.
* **Correction for Bias:** The $\lambda$ term ensures that the $\boldsymbol{\beta}$ and $\boldsymbol{\theta}$ coefficients are **unbiased and efficient** by clearing the residuals of spatial autocorrelation.
* **Stability:** Crucially, the SLX-Error model **does not include the endogenous feedback loop ($\rho \mathbf{W Y}$)**. This makes it a structurally stable model, avoiding the instability issues (like $\rho > 1$) that you encountered with the General Moments SDM.

The SLX-Error model is often used when diagnostics strongly point to both **error dependence** and **exogenous spillovers**, and when the endogenous feedback ($\rho \mathbf{W Y}$) is found to be either insignificant or unstable.

In [97]:
from pysal.model.spreg import GM_Error_Het

SLXE_model = GM_Error_Het(**argsRegre_W,
                          slx_lags=1              # Adds the W*X terms
                         )

SLXE_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,13.17256,1.951204,6.750992,0.0
1,Sanitation_deficit_pct,0.734482,0.015613,47.04185,0.0
2,Education_deficit_pct,0.210654,0.02202,9.566289,0.0
3,W_Sanitation_deficit_pct,-0.084732,0.034307,-2.469789,0.013519
4,W_Education_deficit_pct,0.100745,0.03814,2.641454,0.008255
5,lambda,0.770436,0.019314,39.890691,0.0


What do we have now?

| Variable       | Interpretation                                                                                                                                                                                                                                                                                                                                             |
| :------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **λ (lambda)** | **Strong Spatial Error Dependence:** λ = 0.770 is highly significant, indicating that unobserved shocks are strongly spatially correlated across districts. Disturbances in one area propagate with roughly 77% of their strength to neighboring districts through latent regional forces (e.g., shared policies, institutions, environmental conditions). |

**Core Direct Effects ($X\boldsymbol{\beta}$)**

| Variable                   | Interpretation                                                                                                                                                                           |
| :------------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Sanitation_deficit_pct** | **Dominant Local Driver:** A 1-unit increase in a district’s sanitation deficit increases its Basic Needs Deficit by 0.73 units, confirming sanitation as the strongest local predictor. |
| **Education_deficit_pct**  | **Significant Local Effect:** A 1-unit increase in a district’s education deficit raises its Basic Needs Deficit by 0.21 units.                                                          |

**Exogenous Spillover Effects ($WX\boldsymbol{\theta}$)**

| Variable                     | Interpretation                                                                                                                                                                                                   |
| :--------------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **W_Sanitation_deficit_pct** | **Negative Spillover (Rivalry):** Higher sanitation deficits in neighboring districts are associated with a **lower** local Basic Needs Deficit, suggesting competitive or compensatory allocation of resources. |
| **W_Education_deficit_pct**  | **Positive Spillover (Contagion):** Higher education deficits among neighbors increase local deprivation, indicating shared regional vulnerability through education systems or social conditions.               |

**Synthesis**

> The SLX–Error model demonstrates that the spatial dynamics of Basic Needs Deficit are driven by two complementary mechanisms: (1) **strong structural clustering** (λ) through unobserved, spatially correlated forces, and (2) **contrasting exogenous spillovers** (WX) where sanitation exhibits a rivalry effect and education exhibits a contagion effect. Together, these findings underscore the importance of accounting for both latent regional factors and observable neighbor effects in understanding deprivation.

These results highlight that addressing Basic Needs Deficit in "a distrito" requires a **spatially coordinated approach**. Policies targeting sanitation should consider potential resource competition across neighboring districts, while interventions in education must account for **diffusive effects**, as deprivation in one district can exacerbate outcomes in its neighbors. Ignoring either the latent regional clustering or the observable spillovers could lead to ineffective or unevenly distributed interventions.

## Spatial Regression: SAC


The SAC model combines **two spatial processes**:

$$
y = \rho W y + X\beta + u, \quad u = \lambda W u + \varepsilon
$$

* $\rho$ → spatial lag of the dependent variable (W y)
* $\lambda$ → spatial error (W u)
* $\beta$ → coefficients on observed covariates (X)

We can compute it like this:


In [98]:
sac_model =spreg.GM_Combo_Het(**argsRegre,
    w=peru_perim_graph,
    name_w='peru_perim_graph')

sac_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,15.738191,2.10113,7.490347,0.0
1,Sanitation_deficit_pct,0.741984,0.015887,46.704395,0.0
2,Education_deficit_pct,0.211046,0.021525,9.80487,0.0
3,W_BasicNeeds_deficit_pct,0.009839,0.044581,0.220689,0.825335
4,lambda,0.8027,0.024035,33.397797,0.0


We see that:

| Variable                                             | Interpretation                                                                                                                                                                                                        |
| :--------------------------------------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Sanitation_deficit_pct (β = 0.742, p < 0.001)**    | **Dominant direct effect:** Local sanitation deficits strongly increase Basic Needs Deficit.                                                                                                                          |
| **Education_deficit_pct (β = 0.211, p < 0.001)**     | **Significant direct effect:** Local education deficits also increase Basic Needs Deficit.                                                                                                                            |
| **W_BasicNeeds_deficit_pct (ρ = 0.0098, p = 0.825)** | **Insignificant endogenous feedback:** Neighboring districts’ Basic Needs Deficit has **no measurable direct effect** on the local outcome. The dependent variable does not “diffuse” across neighbors in this model. |
| **λ (lambda = 0.803, p < 0.001)**                    | **Strong spatial error dependence:** Unobserved shocks remain highly correlated across districts, indicating that latent regional factors drive much of the spatial structure.                                        |
Then,

> In the SAC model, the direct impact of local sanitation and education deficits remains robust. The spatial lag of the dependent variable (ρ) is essentially zero and non-significant, meaning that **deprivation does not propagate directly from one district to its neighbors**. Instead, the spatial autocorrelation is almost entirely in the **error term (λ = 0.803)**, reflecting **latent, regionally clustered factors** (e.g., shared policies, infrastructure, or environmental shocks) that are not captured by the observed covariates.

## Spatial Regression: SDM

The Spatial Durbin Model (SDM) is a model that combines the SAR structure with the **spatial lag of the independent variables** (the $WX\theta$ term).

* **When to Use It (Diagnostics):** When both the **Robust LM (error)** and the **Robust LM WX test** are significant (as they are in our case).
* **Model Structure:**
    $$Y = \rho W Y + X\beta + W X \theta + \epsilon$$
    * $\rho W Y$: Neighbors' deficit influences your deficit.
    * $W X \theta$: Neighbors' **independent variables** (Sanitation and Education deficits) influence your deficit.
* **Interpretation:** This model provides the **most complete picture of spillovers.** For example, not only does your neighbor's high Basic Needs Deficit affect yours ($\rho$), but your neighbor's high Education Deficit might also directly contribute to yours ($\theta$).
* **SDM's Advantage:** If the true data generating process is an SDM, omitting the $WX$ term leads to an omitted variable bias in the SAR model. The SDM avoids this risk, making it the **safest starting point**. 

---

In [100]:
sdm_model = spreg.GM_Lag(**argsRegre_W,
    slx_lags=1, robust="white"                          
)
sdm_model.output

,var_names,coefficients,std_err,zt_stat,prob
0,CONSTANT,0.666529,1.159695,0.574745,0.565463
1,Sanitation_deficit_pct,0.761842,0.016511,46.141438,0.0
2,Education_deficit_pct,0.175288,0.021919,7.997127,0.0
3,W_Sanitation_deficit_pct,-0.778413,0.090416,-8.60927,0.0
4,W_Education_deficit_pct,-0.203502,0.063704,-3.194505,0.001401
5,W_BasicNeeds_deficit_pct,1.016223,0.132071,7.694527,0.0


This is what we have:

**Endogenous Spatial Lag**

| Variable                                            | Interpretation                                                                                                                                                                    |
| :-------------------------------------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **W_BasicNeeds_deficit_pct (ρ = 1.016, p < 0.001)** | **Strong endogenous feedback:** A higher Basic Needs Deficit in neighboring districts strongly increases the local district’s deficit, reflecting outcome diffusion or contagion. |

---

**Core Direct Effects ($X\beta$)**

| Variable                   | Interpretation                                                                                                                                                           |
| :------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Sanitation_deficit_pct** | **Dominant local driver:** A 1-unit increase in the district’s own sanitation deficit raises its Basic Needs Deficit by 0.76 units, controlling for all spatial effects. |
| **Education_deficit_pct**  | **Significant local effect:** A 1-unit increase in the district’s own education deficit raises its Basic Needs Deficit by 0.18 units.                                    |



**Exogenous Spillover Effects ($W X\theta$)**

| Variable                     | Interpretation                                                                                                                                                                                                                              |
| :--------------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **W_Sanitation_deficit_pct** | **Negative spillover (rivalry/compensation):** Higher sanitation deficits in neighboring districts **reduce** local Basic Needs Deficit, suggesting resources may be redirected to worse-off neighbors or compensatory dynamics.            |
| **W_Education_deficit_pct**  | **Negative spillover (contagion moderation):** Higher education deficits in neighboring districts **slightly reduce** local deprivation, indicating some complex interaction, possibly due to shared educational interventions or policies. |

The SDM results reveal that Basic Needs Deficit is shaped by **multiple spatial channels**:

> 1. **Strong endogenous diffusion (ρ ≈ 1.02):** neighboring outcomes strongly influence local deprivation, indicating inter-district contagion.
> 2. **Local deficits (β):** sanitation and education remain significant local drivers.
> 3. **Contrasting exogenous spillovers (WX):** neighbor sanitation and education deficits have negative effects on local deprivation, suggesting compensatory dynamics or policy targeting.
>
> This underscores that both **outcome feedback** and **neighbor covariate effects** are crucial to understanding spatial patterns of deprivation, making SDM the most comprehensive model among those estimated.


* **Can we omit the exogenous control on WX?**

In [102]:
# Common Factor Hypothesis test
sdm_model.cfh_test[1]>0.05

np.True_

In [105]:
combo_model=spreg.GM_Combo_Het(**argsRegre,
    w=peru_perim_graph,slx_lags=1,
    name_w='peru_perim_graph')
print(combo_model.summary)

REGRESSION RESULTS
------------------

SUMMARY OF OUTPUT: SPATIALLY WEIGHTED 2SLS- GM-COMBO MODEL (HET) WITH SLX (GNSM)
--------------------------------------------------------------------------------
Data set            :      DisMap
Weights matrix      :peru_perim_graph
Dependent Variable  :BasicNeeds_deficit_pct                Number of Observations:        1871
Mean dependent var  :     38.5867                Number of Variables   :           6
S.D. dependent var  :     18.2951                Degrees of Freedom    :        1865
Pseudo R-squared    :      0.8026
Spatial Pseudo R-squared:  0.1315
N. of iterations    :           1                Step1c computed       :          No

------------------------------------------------------------------------------------
            Variable     Coefficient       Std.Error     z-Statistic     Probability
------------------------------------------------------------------------------------
            CONSTANT         0.86543         2.30228 

In [106]:
# List all variables ending with "_model"
model_list = [name for name in globals() if name.endswith('_model')]
print(model_list)


['RKMeans_model', 'azp_model', 'skater_model', 'ols_model', 'SAR_model', 'SEM_model', 'SLXE_model', 'sdm_model', 'slx_model', 'sac_model', 'combo_model']


In [114]:
# List of your model variable names
model_names = ['ols_model', 'SAR_model', 'SEM_model', 'SLXE_model', 
               'sdm_model', 'slx_model', 'sac_model', 'combo_model']

# Prepare a summary list
summary = []

for name in model_names:
    model = globals()[name]  # get the model object
    # Get AIC if available
    aic = getattr(model, 'aic', None)
    # Get pseudo R2
    r2 = getattr(model, 'r2', None)
    # Get spatial pseudo R2 if available
    pr2 = getattr(model, 'pr2', None)
    summary.append({
        'Model': name,
        'AIC': aic,
        'R2': r2,
        'Spatial Pseudo R2': pr2
    })

# Convert to DataFrame for nice display
import pandas as pd
summary_df = pd.DataFrame(summary)
print(summary_df)


         Model           AIC        R2  Spatial Pseudo R2
0    ols_model  14775.887908  0.530693                NaN
1    SAR_model           NaN       NaN           0.567819
2    SEM_model           NaN       NaN           0.512031
3   SLXE_model           NaN       NaN           0.535807
4    sdm_model           NaN       NaN           0.806578
5    slx_model  14714.263135  0.546868                NaN
6    sac_model           NaN       NaN           0.519571
7  combo_model           NaN       NaN           0.802604



* Combo includes:

  * **Direct local effects (β)** → easy to interpret
  * **Exogenous neighbor effects (WX → θ)** → observable spillovers
  * **Endogenous outcome diffusion (ρ)** → neighbors’ outcomes influence local outcome
  * **Latent spatial error (λ)** → unobserved clustering

* The **problem**: all four effects operate simultaneously and partially absorb each other. For example:

  * WX effects may **change sign or magnitude** once ρ is included.
  * λ is smaller because some clustering is now captured by ρ + WX.

* Result: it becomes **harder to say clearly “a 1-unit increase in neighbor’s deficit raises local deficit by X”** without qualifying that part of it is mediated through endogenous ρ.

---

**Policy interpretation requires clarity of causal channels**

* Policy recommendations usually focus on:

  1. Local interventions → β effects
  2. Neighbor effects → θ effects (contagion vs. rivalry)
  3. Structural clustering → λ (less actionable, because unobserved)

* Combo **mixes all channels**. If a policymaker asks:

  > “Should we invest in sanitation locally, or intervene in neighboring districts?”

  * You now have to **decompose direct β, θ, and ρ effects** to answer — much more complicated than in **SLX–Error**, where:

    * β = local effects
    * θ = exogenous spillovers
    * λ = latent clustering (actionable only as awareness of hidden regional shocks)

---

3. **β coefficients remain interpretable, but θ + ρ interactions are tricky**

* E.g., in your Combo model:

  * W_Sanitation = –0.796 (negative)
  * W_BasicNeeds (ρ) = 0.9 (positive)

* Both effects act on neighbors, but they **partially offset each other**. A naïve policy interpretation could be misleading without decomposition.

---

### ✅ Summary

| Model         | Interpretability                                   | Policy relevance                                                                    |
| ------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------- |
| **SLX–Error** | High → clear β (local) + θ (neighbor) + λ (latent) | Easy to say which interventions matter locally or via neighbors                     |
| **Combo**     | Moderate → β clear, but θ + ρ interact             | Harder to give simple recommendations because endogenous diffusion masks spillovers |

**Bottom line:**

> **Combo is statistically “fullest”, but SLX–Error is more transparent for policy because you can separate local, neighbor, and latent effects clearly.**

